File: scripts/test_vector_search.py
#!/usr/bin/env python3
"""
Test Vector Search Script

This script tests the vector search functionality in the database to verify
that the embeddings are working correctly.
"""

import os
import sys
import argparse
from pathlib import Path
from dotenv import load_dotenv
import openai
from supabase import create_client, Client

# Configure basic logging
import logging
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)

# Constants
OPENAI_MODEL = "text-embedding-ada-002"
LIMIT = 5  # Number of results to return

def get_args():
    """Parse command-line arguments."""
    parser = argparse.ArgumentParser(description="Test Vector Search")
    parser.add_argument(
        "query", 
        help="Search query to test (e.g., 'fruity chocolate coffee')"
    )
    parser.add_argument(
        "--limit", 
        type=int, 
        default=LIMIT,
        help=f"Number of results to return (default: {LIMIT})"
    )
    parser.add_argument(
        "--verbose", 
        action="store_true",
        help="Enable verbose logging"
    )
    return parser.parse_args()

def initialize_clients():
    """Initialize Supabase and OpenAI clients."""
    # Load environment variables from .env file
    env_path = Path(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))) / '.env'
    load_dotenv(dotenv_path=env_path)
    
    # Get credentials from environment
    supabase_url = os.environ.get("SUPABASE_URL")
    supabase_key = os.environ.get("SUPABASE_KEY")
    openai_key = os.environ.get("OPENAI_API_KEY")
    
    # Validate credentials
    if not supabase_url or not supabase_key:
        logger.error("Supabase URL and key are required (set in .env file)")
        sys.exit(1)
    
    has_openai = bool(openai_key)
    if has_openai:
        openai.api_key = openai_key
        logger.info("OpenAI API key configured")
    else:
        logger.warning("No OpenAI API key provided - vector search will fail")
        sys.exit(1)
    
    # Initialize Supabase client
    try:
        supabase = create_client(supabase_url, supabase_key)
        logger.info("Supabase client initialized successfully")
        return supabase
    except Exception as e:
        logger.error(f"Failed to initialize Supabase client: {e}")
        sys.exit(1)

def generate_embedding(query_text):
    """Generate an embedding for the search query."""
    try:
        logger.info(f"Generating embedding for query: {query_text}")
        client = openai.OpenAI(api_key=os.environ.get("OPENAI_API_KEY"))
        response = client.embeddings.create(
            model=OPENAI_MODEL,
            input=query_text
        )
        
        embedding = response.data[0].embedding
        logger.info(f"Generated embedding with {len(embedding)} dimensions")
        return embedding
    except Exception as e:
        logger.error(f"Error generating embedding: {e}")
        sys.exit(1)

def perform_vector_search(supabase, embedding, limit):
    """Perform vector search in the database."""
    try:
        # Convert embedding to string format for RPC call
        embedding_str = f"[{','.join(str(float(x)) for x in embedding)}]"
        
        # Call the RPC function for vector search
        logger.info(f"Performing vector search with limit {limit}")
        
        # Add debug info about the request
        logger.info(f"Using supabase URL: {os.environ.get('SUPABASE_URL')}")
        logger.info(f"Embedding dimensions: {len(embedding)}")
        
        # Perform the RPC call with detailed error handling
        try:
            result = supabase.rpc(
                "search_coffee_by_flavor_vector",
                {
                    "query_embedding": embedding_str,
                    "match_threshold": 0.5,
                    "match_count": limit
                }
            ).execute()
            
            if result.data:
                logger.info(f"Search successful, found {len(result.data)} results")
                return result.data
            else:
                logger.warning("No search results found")
                return []
        except Exception as rpc_error:
            logger.error(f"RPC call failed: {rpc_error}")
            if hasattr(rpc_error, 'details'):
                logger.error(f"Error details: {rpc_error.details}")
            raise rpc_error
    except Exception as e:
        logger.error(f"Error performing vector search: {e}")
        # Check if we can get more details about the error
        if hasattr(e, '__dict__'):
            for attr, value in e.__dict__.items():
                logger.error(f"Error attribute {attr}: {value}")
        return []

def main():
    """Main entry point."""
    args = get_args()
    
    if args.verbose:
        logger.setLevel(logging.DEBUG)
    
    logger.info("Starting vector search test")
    
    # Initialize clients
    supabase = initialize_clients()
    
    # Generate embedding for query
    embedding = generate_embedding(args.query)
    
    # Perform vector search
    results = perform_vector_search(supabase, embedding, args.limit)
    
    # Display results
    print("\n=== Vector Search Results ===")
    if not results:
        print("No results found")
    else:
        print(f"Found {len(results)} results:")
        for i, coffee in enumerate(results):
            # Print all available fields
            print(f"\n{i+1}. {coffee.get('name', 'Unknown')} (Similarity: {coffee.get('similarity', 0):.4f})")
            print(f"   ID: {coffee.get('id', 'Unknown')}")
            print(f"   Roaster: {coffee.get('roaster_name', 'Unknown')}")
            print(f"   Flavors: {', '.join(coffee.get('flavor_tags', []))}")
            print(f"   Roast Level: {coffee.get('roast_level', 'Unknown')}")
            
            if coffee.get('process_method'):
                print(f"   Process Method: {coffee['process_method']}")
                
            if coffee.get('is_featured'):
                print(f"   Featured: Yes")
                
            if coffee.get('price'):
                print(f"   Price: ${coffee['price']}")
                
            # Print all fields for debugging
            print("\n   All available fields:")
            for key, value in coffee.items():
                if key not in ['id', 'name', 'roaster_name', 'flavor_tags', 'roast_level', 'process_method', 'is_featured', 'price', 'similarity']:
                    print(f"      {key}: {value}")


if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        logger.info("Test interrupted by user")
        sys.exit(1)
    except Exception as e:
        logger.error(f"Unhandled error: {e}")
        sys.exit(1)

File: debug/archive/test_vector_search.js
/**
 * Consolidated test script for the vector search functionality
 * Tests both direct database access and MCP endpoint
 */
import { createClient } from '@supabase/supabase-js';
import OpenAI from 'openai';
import dotenv from 'dotenv';
import fetch from 'node-fetch';

// Load environment variables from .env file
dotenv.config({ path: '../.env' });

const supabaseUrl = process.env.SUPABASE_URL;
const supabaseKey = process.env.SUPABASE_KEY;
const openaiApiKey = process.env.OPENAI_API_KEY;
const mcpServerUrl = 'https://better-beans-mcp-server.al-ricotta.workers.dev/rpc';

if (!supabaseUrl || !supabaseKey) {
  console.error('Missing SUPABASE_URL or SUPABASE_KEY');
  process.exit(1);
}

if (!openaiApiKey) {
  console.error('Missing OPENAI_API_KEY');
  process.exit(1);
}

console.log(`Using Supabase URL: ${supabaseUrl}`);
console.log(`Supabase Key: ${supabaseKey.substring(0, 5)}...${supabaseKey.substring(supabaseKey.length - 5)}`);
console.log(`OpenAI Key: ${openaiApiKey.substring(0, 5)}...${openaiApiKey.substring(openaiApiKey.length - 5)}`);

const supabase = createClient(supabaseUrl, supabaseKey);
const openai = new OpenAI({
  apiKey: openaiApiKey
});

// Test the MCP endpoint
async function testMcpEndpoint(flavorProfile) {
  console.log(`\n🌐 Testing MCP endpoint with "${flavorProfile.join(', ')}" flavor profile...`);
  
  try {
    const response = await fetch(mcpServerUrl, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        jsonrpc: '2.0',
        id: 'test-similarity-search',
        method: 'similarity_search',
        params: {
          flavorProfile,
          maxResults: 5
        }
      })
    });
    
    const result = await response.json();
    
    if (result.error) {
      console.error(`❌ MCP endpoint error: ${JSON.stringify(result.error)}`);
      return;
    }
    
    // Parse the content from the MCP response
    let searchResults = {};
    try {
      if (result.result && result.result.content && result.result.content[0] && result.result.content[0].text) {
        searchResults = JSON.parse(result.result.content[0].text);
      } else {
        console.error('❌ Unexpected MCP response format:', JSON.stringify(result));
        return;
      }
    } catch (error) {
      console.error('❌ Error parsing MCP response:', error);
      return;
    }
    
    console.log(`✅ Found ${searchResults.results ? searchResults.results.length : 0} results via MCP endpoint`);
    
    if (searchResults.results && searchResults.results.length > 0) {
      console.log('\n   Top results:');
      searchResults.results.forEach((result, i) => {
        console.log(`   ${i+1}. ${result.coffee.name}`);
        console.log(`      Flavor tags: ${JSON.stringify(result.coffee.flavorTags)}`);
        console.log(`      Similarity: ${result.similarityScore?.toFixed(4) || 'N/A'}`);
        if (result.matchingTags && result.matchingTags.length > 0) {
          console.log(`      Matching tags: ${JSON.stringify(result.matchingTags)}`);
        }
      });
    } else {
      console.log('   No results returned');
    }
  } catch (e) {
    console.error('❌ Error testing MCP endpoint:', e);
  }
}

// Generate an embedding using OpenAI
async function generateEmbedding(text) {
  console.log(`Generating embedding for: "${text}"`);
  
  try {
    const response = await openai.embeddings.create({
      model: "text-embedding-3-small",
      input: text,
      encoding_format: "float"
    });
    
    const embedding = response.data[0].embedding;
    console.log(`✅ Generated ${embedding.length}-dimensional embedding`);
    return embedding;
  } catch (error) {
    console.error('❌ Error generating embedding:', error);
    throw error;
  }
}

// Test vector search with real embedding
async function testDirectVectorSearch(flavorProfile) {
  console.log(`\n🔍 Testing direct vector search with "${flavorProfile.join(', ')}" flavor profile...`);
  
  try {
    // Generate embedding for the flavor profile
    const embedding = await generateEmbedding(flavorProfile.join(', '));
    
    // Fixed threshold that works with our database
    const threshold = 0.01;
    console.log(`📊 Testing with threshold: ${threshold}`);
    
    const { data, error } = await supabase
      .rpc('search_coffee_by_flavor_vector', {
        query_embedding: `[${embedding.join(',')}]`,
        match_threshold: threshold,
        match_count: 5,
        match_offset: 0
      });
      
    if (error) {
      console.error(`❌ ERROR with direct vector search:`, error);
    } else {
      console.log(`✅ Found ${data?.length || 0} results with direct vector search`);
      
      if (data && data.length > 0) {
        console.log('\n   Top results:');
        data.forEach((result, i) => {
          console.log(`   ${i+1}. ${result.name} (similarity: ${result.similarity.toFixed(4)}, distance: ${result.distance.toFixed(4)})`);
          console.log(`      Flavor tags: ${JSON.stringify(result.flavor_tags)}`);
        });
      } else {
        console.log('   No results returned');
      }
    }
  } catch (e) {
    console.error('❌ EXCEPTION in direct vector search:', e);
  }
}

// Run all tests
async function runAllTests() {
  try {
    const flavorProfiles = [
      ['berry', 'fruit', 'clean'],
      ['chocolate', 'nutty'],
      ['citrus', 'floral', 'bright']
    ];
    
    for (const profile of flavorProfiles) {
      await testMcpEndpoint(profile);
      await testDirectVectorSearch(profile);
      console.log('\n' + '-'.repeat(80) + '\n');
    }
    
    console.log('✨ All tests completed');
  } catch (e) {
    console.error('❌ Test execution failed:', e);
  }
}

// Execute all tests
runAllTests();
